<b>Instructions</b>

<h1>Initiation</h1>

<p>Your input should be a haplotype file of the following form:</p>

<table>
  <tr>
    <th>sid</th>
    <th>rs0000001_A</th>
    <th>rs0000002_G</th>
    <th>rs0000003_C</th>
    <th>...</th>
  </tr>
  <tr>
    <th>0001</th>
    <td>0</td>
    <td>0</td>
    <td>0</td>
    <td>...</td>
  </tr>
  <tr>
    <th>0001</th>
    <td>0</td>
    <td>1</td>
    <td>0</td>
    <td>...</td>
  </tr>
  <tr>
    <th>0002</th>
    <td>1</td>
    <td>0</td>
    <td>0</td>
    <td>...</td>
  </tr>
  <tr>
    <th>0002</th>
    <td>1</td>
    <td>0</td>
    <td>0</td>
    <td>...</td>
  </tr>
 </table>
 
 <p>Each row is chromosome from subject sid; note that each has two chromosomes with the same id.  Alternate alleles of SNP </kbd>rsid_Alt</kbd> are indicated with a 1 and reference alleles with a 0.  Transpose the file with one of the following commands.  Note that the sid row is purposely excluded.</p>
 
 <pre>
 <code>
 n_fields=$(awk 'END{print NF}' file)
 for i in `seq 2 $n_fields`
 do
    cut -f$i file | paste -s # paste column i as a row
 done > transpose_file
 </code>
 </pre>
 
 or
 
 <pre>
 <code>
 awk 'BEGIN{OFS="\t"}; {for(j=2;j<=NF;j++) {a[NR,j]=$j; n_rows=NR; n_cols=(n_cols<NF?NF:n_cols)} } END{for (j=2;j<=n_cols;j++) {for (i=1;i<=n_rows;i++) {printf "%s%s",a[i,j],(i==n_rows?"\n":"\t")} } }' file > transpose file
 </code>
 </pre>
 
 The output will be something like:
 
 <table>
  <tr>
    <th>rs0000001_A</th>
    <td>0</td>
    <td>0</td>
    <td>1</td>
    <td>1</td>
    <td>...</td>
  </tr>
  <tr>
    <th>rs0000002_G</th>
    <td>0</td>
    <td>1</td>
    <td>0</td>
    <td>0</td>
    <td>...</td>
  </tr>
  <tr>
    <th>rs0000003_C</th>
    <td>0</td>
    <td>0</td>
    <td>0</td>
    <td>0</td>
    <td>...</td>
  </tr>
 </table>
 
 <p>The initiation step overlaps columns of <kbd>transpose_file</kbd>.  For users not running on an HPC, the <kbd>ChromosomeOverlap_initiation_sub.sh</kbd> program loops through all <i>variable</i> chromosomes and overlaps them with each column <var>j</var> of the transposed matrix such that <var>j>i<sub>&sigma-1</sub>.  The intermediate files generated by <kbd>ChromosomeOverlap_initiation.sh</kbd> look like </p>
  
  <table>
    <tr>
      <td>10</td>
      <td>00</td>
      <td>01,02,04,07</td>
    </tr>
    <tr>
      <td>10</td>
      <td>10</td>
      <td>03,05</td>
    </tr>
    <tr>
      <td>10</td>
      <td>01</td>
      <td>06,08</td>
    </tr>
    <tr>
      <td>10</td>
      <td>11</td>
      <td>09</td>
    </tr>
   </table>
    
<p>The first column is the number of the variable chromosome (i.e., the tenth column of the matrix, including the column of rsids).  The second column is the "pattern type," which indicates the allele combination when a single fixed chromosome is placed next to the fixed one.  The third column is the list of SNPs (starting from 01) at which the indiciated pattern is observed.  Chromosome Overlap will produce a <i>meta-chromosome</i> <kbd>01_0,02_0,04_0,05_0,09_1</kbd> indicating SNPs 1, 2, 4, and 5 are all in the REF state and SNP 9 is in the ALT state. 
